### PR TITLE
Add evalMaybe, evalMaybeM and evalEitherM

### LINKED ISSUE #368: Add evalMaybeM :: MonadTest m => m (Maybe a) -> m a
As the issue title says, I suggest we add this. I would also suggest `evalMaybe` and `evalEitherM`. I generally use pattern matching for these, but then I don't get nice highlighting :cry:

### PR DESCRIPTION
**Pull request by markus1189**
adds:

```haskell
evalMaybe :: (MonadTest m, Show a, HasCallStack) => Maybe a -> m a
evalMaybeM :: (MonadTest m, Show a, MonadCatch m, HasCallStack) => m (Maybe a) -> m a
evalEitherM :: (MonadTest m, Show x, MonadCatch m, HasCallStack) => m (Either x a) -> m a
```

Fixes #368

### CODE REVIEW

**Comment by kirelagin** [comment]
Btw, I guess, `Maybe` could be covered by a more generic case, something like:

```haskell
evalAlternative :: (..., Alternative f, Eq (f a)) => f a -> m a
```

**Comment by markus1189** [comment]
@kirelagin Interesting thought, but how would you access the concrete value inside `f a` if it's not the `empty` case?

**Comment by kirelagin** [comment]
Oh... You are right, sorry for the confusion 🙈.

Anyway, I came to this PR for `evalMaybe`, which I have been reimplementing in test for various my packages for more than 2 years now, and when I needed it yet again, I thought I’d finally open a PR, and found this. So, I’m, obviously, 👍 on this and let’s get this merged :).

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (kirelagin) [inline hedgehog/src/Hedgehog/Internal/Property.hs:832]:
```suggestion
--   'Either' is 'Left'. Otherwise returns the value in the 'Right'.
```

Nowadays, people usually don’t put two spaces after punctuation. (And, most importantly, other comments in this repository have only one space, as far as I can tell.)

**Comment by markus1189** [comment]
No worries and thanks for taking a look :). Ooops about the two spaces,
it's become a habit at some point... Will change to 1 space

**Comment by EarthCitizen** [comment]
@kirelagin @markus1189 Is this PR abandoned?

**Comment by markus1189** [comment]
There never was any sign whether this has a chance of being merged, so before investing any more time, it would be good to know if there is any interest at all?

**Comment by ocharles** [comment]
I am not a maintainer, but I would love to have these. The more general one would take `Foldable` and try and safely `head` the `toList` of it, but I'd rather just have these explicit ones for now.

**Comment by EarthCitizen** [comment]
@markus1189 I cannot speak to the way they are implemented, nor if there are needed improvements, but I find them useful. I was actually looking for an `evalMaybe` when I stumbled upon this PR.

**Comment by sshine** [comment]
I also came looking for `evalMaybe`. I look forward to seeing this addressed. :)

File: hedgehog/src/Hedgehog.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog.hs:116]: It'd be nice to have those in alphabetical order, after `evalEitherXyz`.

File: hedgehog/src/Hedgehog.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog.hs:171]: It'd be nice to have those in alphabetical order.

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:81]: It'd be nice to have those in alphabetical order, after `evalEitherXyz`.

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:819]:
```suggestion
evalMaybeM =
  evalMaybe <=< evalM
```

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:835]:
```suggestion
evalEitherM =
  evalEither <=< evalM
```

**Review by moodmosaic — COMMENTED** [review]
Thanks for the pull request, @markus1189 :+1:

We apologize for the delay. I just left a few minor comments that'd be nice to have before we merge.

Also, this branch has conflicts that must be resolved
* hedgehog/src/Hedgehog.hs
* hedgehog/src/Hedgehog/Internal/Property.hs

Let me know if you want me to take care of those (changes and conflicts) as this has been opened for a while now.

File: hedgehog/src/Hedgehog.hs
Comment (markus1189) [reply hedgehog/src/Hedgehog.hs:116]: sorted evalMaybeXyz after evalEither

File: hedgehog/src/Hedgehog.hs
Comment (markus1189) [reply hedgehog/src/Hedgehog.hs:171]: sorted evalMaybeXyz after evalEither

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (markus1189) [reply hedgehog/src/Hedgehog/Internal/Property.hs:81]: sorted evalMaybeXyz after evalEither

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (markus1189) [reply hedgehog/src/Hedgehog/Internal/Property.hs:819]: done

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (markus1189) [reply hedgehog/src/Hedgehog/Internal/Property.hs:835]: done

**Comment by markus1189** [comment]
@moodmosaic good to see this being picked up after all.  I think the issues from your first review are adressed.

I liked @ocharles suggestion of generalizing it to `Foldable` and checking the `head`, so while I have your attention (:grin:) I also added `evalHead` and `evalHeadM` that works for any `Foldable`, but can't give as precise error messages.

**Comment by moodmosaic** [comment]
@markus1189, thank you :+1: Would you mind moving https://github.com/hedgehogqa/haskell-hedgehog/pull/381/commits/dd579c5da167e9c283240e2f2eff7a383dd540d1 on its own branch/PR? I'm about to plan on cutting a release after merging this, and then review the foldable one separately.

**Comment by markus1189** [comment]
@moodmosaic Done!

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:857]:
```suggestion
    withFrozenCallStack $ failWith Nothing "the value was Nothing"
```

**Review by moodmosaic — APPROVED** [review]
:+1:

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:861]:
```suggestion
-- | Fails the test if the action throws an exception, or if the
```

File: hedgehog/src/Hedgehog/Internal/Property.hs
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Property.hs:862]:
```suggestion
--   'Maybe' is 'Nothing', otherwise returns the value in the 'Just'.
```

**Review by moodmosaic — APPROVED** [review]
:+1: I've run a quick test locally, it looks good.

```diff
diff --git a/hedgehog-example/src/Test/Example/STLC.hs b/hedgehog-example/src/Test/Example/STLC.hs
index a3889b2..4b97853 100644
--- a/hedgehog-example/src/Test/Example/STLC.hs
+++ b/hedgehog-example/src/Test/Example/STLC.hs
@@ -286,6 +286,19 @@ prop_illtyped =
     _t <- evalEither (typecheck ex)
     success
 
+whenLeft :: Applicative m => Either a b -> (a -> m ()) -> m ()
+whenLeft (Left x) f =
+  f x
+whenLeft _ _ =
+  pure ()
+
+prop_illtyped_maybe :: Property
+prop_illtyped_maybe =
+  property $ do
+    ex <- forAll genIllTypedExpr
+    _t <- evalMaybe (whenLeft (typecheck ex) (\_ -> Nothing))
+    success
+
 prop_consistent :: Property
 prop_consistent =
   property $ do
```

```
λ check prop_illtyped_maybe
  ✗ <interactive> failed at /src/haskell-hedgehog/hedgehog-example/src/Test/Example/STLC.hs:299:11
    after 1 test and 5 shrinks.

        ┏━━ /src/haskell-hedgehog/hedgehog-example/src/Test/Example/STLC.hs ━━━
    295 ┃ prop_illtyped_maybe :: Property
    296 ┃ prop_illtyped_maybe =
    297 ┃   property $ do
    298 ┃     ex <- forAll genIllTypedExpr
        ┃     │ EApp (ELam "a" TBool (EBool True)) (EInt 0)
    299 ┃     _t <- evalMaybe (whenLeft (typecheck ex) (\_ -> Nothing))
        ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        ┃     │ the value was Nothing
    300 ┃     success

    This failure can be reproduced by running:
    > recheck (Size 0) (Seed 8279953953216453507 1134259921219244333) <property>
```

**Pull request merged**
